File: android/src/main/kotlin/com/affise/attribution/affise_attribution_lib/factories/AffiseSubtypeEventsFactory.kt
package com.affise.attribution.affise_attribution_lib.factories

import com.affise.attribution.events.Event
import com.affise.attribution.events.subscription.*
import com.affise.attribution.affise_attribution_lib.ext.toJSONObject
import org.json.JSONObject

class AffiseSubtypeEventsFactory : AffiseBaseEvensFactory() {

    override fun event(map: Map<*, *>?): Event? {
        return when (map?.get(KEY_SUBTYPE)) {
            SubscriptionParameters.AFFISE_SUB_INITIAL_SUBSCRIPTION -> eventInitialSubscriptionEvent(map)
            SubscriptionParameters.AFFISE_SUB_INITIAL_TRIAL -> eventInitialTrialEvent(map)
            SubscriptionParameters.AFFISE_SUB_INITIAL_OFFER -> eventInitialOfferEvent(map)
            SubscriptionParameters.AFFISE_SUB_FAILED_TRIAL -> eventFailedTrialEvent(map)
            SubscriptionParameters.AFFISE_SUB_FAILED_OFFERISE -> eventFailedOfferiseEvent(map)
            SubscriptionParameters.AFFISE_SUB_FAILED_SUBSCRIPTION -> eventFailedSubscriptionEvent(map)
            SubscriptionParameters.AFFISE_SUB_FAILED_TRIAL_FROM_RETRY -> eventFailedTrialFromRetryEvent(map)
            SubscriptionParameters.AFFISE_SUB_FAILED_OFFER_FROM_RETRY -> eventFailedOfferFromRetryEvent(map)
            SubscriptionParameters.AFFISE_SUB_FAILED_SUBSCRIPTION_FROM_RETRY -> eventFailedSubscriptionFromRetryEvent(map)
            SubscriptionParameters.AFFISE_SUB_TRIAL_IN_RETRY -> eventTrialInRetryEvent(map)
            SubscriptionParameters.AFFISE_SUB_OFFER_IN_RETRY -> eventOfferInRetryEvent(map)
            SubscriptionParameters.AFFISE_SUB_SUBSCRIPTION_IN_RETRY -> eventSubscriptionInRetryEvent(map)
            SubscriptionParameters.AFFISE_SUB_CONVERTED_TRIAL -> eventConvertedTrialEvent(map)
            SubscriptionParameters.AFFISE_SUB_CONVERTED_OFFER -> eventConvertedOfferEvent(map)
            SubscriptionParameters.AFFISE_SUB_REACTIVATED_SUBSCRIPTION -> eventReactivatedSubscriptionEvent(map)
            SubscriptionParameters.AFFISE_SUB_RENEWED_SUBSCRIPTION -> eventRenewedSubscriptionEvent(map)
            SubscriptionParameters.AFFISE_SUB_CONVERTED_TRIAL_FROM_RETRY -> eventConvertedTrialFromRetryEvent(map)
            SubscriptionParameters.AFFISE_SUB_CONVERTED_OFFER_FROM_RETRY -> eventConvertedOfferFromRetryEvent(map)
            SubscriptionParameters.AFFISE_SUB_RENEWED_SUBSCRIPTION_FROM_RETRY -> eventRenewedSubscriptionFromRetryEvent(map)
            else -> null
        }
    }

    private fun getDataJSONObject(map: Map<*, *>): JSONObject {
        return (map[KEY_DATA] as? Map<*, *>)?.toJSONObject() ?: JSONObject()
    }

    private fun eventInitialSubscriptionEvent(map: Map<*, *>): InitialSubscriptionEvent {
        val userData = getUserData(map)
        val data = getDataJSONObject(map)

        return InitialSubscriptionEvent(
            data,
            userData,
        ).apply {
            addPredefinedParameters(this, map)
        }
    }

    private fun eventInitialTrialEvent(map: Map<*, *>): InitialTrialEvent {
        val userData = getUserData(map)
        val data = getDataJSONObject(map)
        
        return InitialTrialEvent(
            data,
            userData,
        ).apply {
            addPredefinedParameters(this, map)
        }
    }

    private fun eventInitialOfferEvent(map: Map<*, *>): InitialOfferEvent {
        val userData = getUserData(map)
        val data = getDataJSONObject(map)
        
        return InitialOfferEvent(
            data,
            userData,
        ).apply {
            addPredefinedParameters(this, map)
        }
    }

    private fun eventFailedTrialEvent(map: Map<*, *>): FailedTrialEvent {
        val userData = getUserData(map)
        val data = getDataJSONObject(map)
        
        return FailedTrialEvent(
            data,
            userData,
        ).apply {
            addPredefinedParameters(this, map)
        }
    }

    private fun eventFailedOfferiseEvent(map: Map<*, *>): FailedOfferiseEvent {
        val userData = getUserData(map)
        val data = getDataJSONObject(map)
        
        return FailedOfferiseEvent(
            data,
            userData,
        ).apply {
            addPredefinedParameters(this, map)
        }
    }

    private fun eventFailedSubscriptionEvent(map: Map<*, *>): FailedSubscriptionEvent {
        val userData = getUserData(map)
        val data = getDataJSONObject(map)
        
        return FailedSubscriptionEvent(
            data,
            userData,
        ).apply {
            addPredefinedParameters(this, map)
        }
    }

    private fun eventFailedTrialFromRetryEvent(map: Map<*, *>): FailedTrialFromRetryEvent {
        val userData = getUserData(map)
        val data = getDataJSONObject(map)
        
        return FailedTrialFromRetryEvent(
            data,
            userData,
        ).apply {
            addPredefinedParameters(this, map)
        }
    }

    private fun eventFailedOfferFromRetryEvent(map: Map<*, *>): FailedOfferFromRetryEvent {
        val userData = getUserData(map)
        val data = getDataJSONObject(map)
        
        return FailedOfferFromRetryEvent(
            data,
            userData,
        ).apply {
            addPredefinedParameters(this, map)
        }
    }

    private fun eventFailedSubscriptionFromRetryEvent(map: Map<*, *>): FailedSubscriptionFromRetryEvent {
        val userData = getUserData(map)
        val data = getDataJSONObject(map)
        
        return FailedSubscriptionFromRetryEvent(
            data,
            userData,
        ).apply {
            addPredefinedParameters(this, map)
        }
    }

    private fun eventTrialInRetryEvent(map: Map<*, *>): TrialInRetryEvent {
        val userData = getUserData(map)
        val data = getDataJSONObject(map)
        
        return TrialInRetryEvent(
            data,
            userData,
        ).apply {
            addPredefinedParameters(this, map)
        }
    }

    private fun eventOfferInRetryEvent(map: Map<*, *>): OfferInRetryEvent {
        val userData = getUserData(map)
        val data = getDataJSONObject(map)
        
        return OfferInRetryEvent(
            data,
            userData,
        ).apply {
            addPredefinedParameters(this, map)
        }
    }

    private fun eventSubscriptionInRetryEvent(map: Map<*, *>): SubscriptionInRetryEvent {
        val userData = getUserData(map)
        val data = getDataJSONObject(map)
        
        return SubscriptionInRetryEvent(
            data,
            userData,
        ).apply {
            addPredefinedParameters(this, map)
        }
    }

    private fun eventConvertedTrialEvent(map: Map<*, *>): ConvertedTrialEvent {
        val userData = getUserData(map)
        val data = getDataJSONObject(map)
        
        return ConvertedTrialEvent(
            data,
            userData,
        ).apply {
            addPredefinedParameters(this, map)
        }
    }

    private fun eventConvertedOfferEvent(map: Map<*, *>): ConvertedOfferEvent {
        val userData = getUserData(map)
        val data = getDataJSONObject(map)
        
        return ConvertedOfferEvent(
            data,
            userData,
        ).apply {
            addPredefinedParameters(this, map)
        }
    }

    private fun eventReactivatedSubscriptionEvent(map: Map<*, *>): ReactivatedSubscriptionEvent {
        val userData = getUserData(map)
        val data = getDataJSONObject(map)
        
        return ReactivatedSubscriptionEvent(
            data,
            userData,
        ).apply {
            addPredefinedParameters(this, map)
        }
    }

    private fun eventRenewedSubscriptionEvent(map: Map<*, *>): RenewedSubscriptionEvent {
        val userData = getUserData(map)
        val data = getDataJSONObject(map)
        
        return RenewedSubscriptionEvent(
            data,
            userData,
        ).apply {
            addPredefinedParameters(this, map)
        }
    }

    private fun eventConvertedTrialFromRetryEvent(map: Map<*, *>): ConvertedTrialFromRetryEvent {
        val userData = getUserData(map)
        val data = getDataJSONObject(map)
        
        return ConvertedTrialFromRetryEvent(
            data,
            userData,
        ).apply {
            addPredefinedParameters(this, map)
        }
    }

    private fun eventConvertedOfferFromRetryEvent(map: Map<*, *>): ConvertedOfferFromRetryEvent {
        val userData = getUserData(map)
        val data = getDataJSONObject(map)
        
        return ConvertedOfferFromRetryEvent(
            data,
            userData,
        ).apply {
            addPredefinedParameters(this, map)
        }
    }

    private fun eventRenewedSubscriptionFromRetryEvent(map: Map<*, *>): RenewedSubscriptionFromRetryEvent {
        val userData = getUserData(map)
        val data = getDataJSONObject(map)
        
        return RenewedSubscriptionFromRetryEvent(
            data,
            userData,
        ).apply {
            addPredefinedParameters(this, map)
        }
    }
}

File: ios/Classes/factories/AffiseSubtypeEventsFactory.swift
import Foundation
import AffiseAttributionLib

internal class AffiseSubtypeEventsFactory : AffiseBaseEvensFactory, AffiseBaseEvens {
    func event(_ map: [String: Any?]?) -> Event? {
        let subtype = map?["subtype"] as? String?
        
        switch subtype {
        case SubscriptionParameters.AFFISE_SUB_INITIAL_SUBSCRIPTION.rawValue: return eventInitialSubscriptionEvent(map)
        case SubscriptionParameters.AFFISE_SUB_INITIAL_TRIAL.rawValue: return eventInitialTrialEvent(map)
        case SubscriptionParameters.AFFISE_SUB_INITIAL_OFFER.rawValue: return eventInitialOfferEvent(map)
        case SubscriptionParameters.AFFISE_SUB_FAILED_TRIAL.rawValue: return eventFailedTrialEvent(map)
        case SubscriptionParameters.AFFISE_SUB_FAILED_OFFERISE.rawValue: return eventFailedOfferiseEvent(map)
        case SubscriptionParameters.AFFISE_SUB_FAILED_SUBSCRIPTION.rawValue: return eventFailedSubscriptionEvent(map)
        case SubscriptionParameters.AFFISE_SUB_FAILED_TRIAL_FROM_RETRY.rawValue: return eventFailedTrialFromRetryEvent(map)
        case SubscriptionParameters.AFFISE_SUB_FAILED_OFFER_FROM_RETRY.rawValue: return eventFailedOfferFromRetryEvent(map)
        case SubscriptionParameters.AFFISE_SUB_FAILED_SUBSCRIPTION_FROM_RETRY.rawValue: return eventFailedSubscriptionFromRetryEvent(map)
        case SubscriptionParameters.AFFISE_SUB_TRIAL_IN_RETRY.rawValue: return eventTrialInRetryEvent(map)
        case SubscriptionParameters.AFFISE_SUB_OFFER_IN_RETRY.rawValue: return eventOfferInRetryEvent(map)
        case SubscriptionParameters.AFFISE_SUB_SUBSCRIPTION_IN_RETRY.rawValue: return eventSubscriptionInRetryEvent(map)
        case SubscriptionParameters.AFFISE_SUB_CONVERTED_TRIAL.rawValue: return eventConvertedTrialEvent(map)
        case SubscriptionParameters.AFFISE_SUB_CONVERTED_OFFER.rawValue: return eventConvertedOfferEvent(map)
        case SubscriptionParameters.AFFISE_SUB_REACTIVATED_SUBSCRIPTION.rawValue: return eventReactivatedSubscriptionEvent(map)
        case SubscriptionParameters.AFFISE_SUB_RENEWED_SUBSCRIPTION.rawValue: return eventRenewedSubscriptionEvent(map)
        case SubscriptionParameters.AFFISE_SUB_CONVERTED_TRIAL_FROM_RETRY.rawValue: return eventConvertedTrialFromRetryEvent(map)
        case SubscriptionParameters.AFFISE_SUB_CONVERTED_OFFER_FROM_RETRY.rawValue: return eventConvertedOfferFromRetryEvent(map)
        case SubscriptionParameters.AFFISE_SUB_RENEWED_SUBSCRIPTION_FROM_RETRY.rawValue: return eventRenewedSubscriptionFromRetryEvent(map)
        case SubscriptionParameters.AFFISE_SUB_UNSUBSCRIPTION.rawValue: return eventUnsubscriptionEvent(map)
        default:
            return nil
        }
    }
   
    private func eventInitialSubscriptionEvent(_ map: [String: Any?]?) -> Event {
        let userData = getUserData(map)
        let data = getDataArray(map)

        var event = InitialSubscriptionEvent(
            data: data,
            userData: userData
        )

       addPredefinedParameters(event: &event, map: map)
       return event
    }

    private func eventInitialTrialEvent(_ map: [String: Any?]?) -> InitialTrialEvent {
        let userData = getUserData(map)
        let data = getDataArray(map)

        var event = InitialTrialEvent(
            data: data,
            userData: userData
        )
        addPredefinedParameters(event: &event, map: map)
        return event
    }

    private func eventInitialOfferEvent(_ map: [String: Any?]?) -> InitialOfferEvent {
        let userData = getUserData(map)
        let data = getDataArray(map)

        var event = InitialOfferEvent(
            data: data,
            userData: userData
        )
        addPredefinedParameters(event: &event, map: map)
        return event
    }

    private func eventFailedTrialEvent(_ map: [String: Any?]?) -> FailedTrialEvent {
        let userData = getUserData(map)
        let data = getDataArray(map)

        var event = FailedTrialEvent(
            data: data,
            userData: userData
        )
        addPredefinedParameters(event: &event, map: map)
        return event
    }

    private func eventFailedOfferiseEvent(_ map: [String: Any?]?) -> FailedOfferiseEvent {
        let userData = getUserData(map)
        let data = getDataArray(map)

        var event = FailedOfferiseEvent(
            data: data,
            userData: userData
        )
        addPredefinedParameters(event: &event, map: map)
        return event
    }

    private func eventFailedSubscriptionEvent(_ map: [String: Any?]?) -> FailedSubscriptionEvent {
        let userData = getUserData(map)
        let data = getDataArray(map)

        var event = FailedSubscriptionEvent(
            data: data,
            userData: userData
        )
        addPredefinedParameters(event: &event, map: map)
        return event
    }

    private func eventFailedTrialFromRetryEvent(_ map: [String: Any?]?) -> FailedTrialFromRetryEvent {
        let userData = getUserData(map)
        let data = getDataArray(map)

        var event = FailedTrialFromRetryEvent(
            data: data,
            userData: userData
        )
        addPredefinedParameters(event: &event, map: map)
        return event
    }

    private func eventFailedOfferFromRetryEvent(_ map: [String: Any?]?) -> FailedOfferFromRetryEvent {
        let userData = getUserData(map)
        let data = getDataArray(map)

        var event = FailedOfferFromRetryEvent(
            data: data,
            userData: userData
        )
        addPredefinedParameters(event: &event, map: map)
        return event
    }

    private func eventFailedSubscriptionFromRetryEvent(_ map: [String: Any?]?) -> FailedSubscriptionFromRetryEvent {
        let userData = getUserData(map)
        let data = getDataArray(map)

        var event = FailedSubscriptionFromRetryEvent(
            data: data,
            userData: userData
        )
        addPredefinedParameters(event: &event, map: map)
        return event
    }

    private func eventTrialInRetryEvent(_ map: [String: Any?]?) -> TrialInRetryEvent {
        let userData = getUserData(map)
        let data = getDataArray(map)

        var event = TrialInRetryEvent(
            data: data,
            userData: userData
        )
        addPredefinedParameters(event: &event, map: map)
        return event
    }

    private func eventOfferInRetryEvent(_ map: [String: Any?]?) -> OfferInRetryEvent {
        let userData = getUserData(map)
        let data = getDataArray(map)

        var event = OfferInRetryEvent(
            data: data,
            userData: userData
        )
        addPredefinedParameters(event: &event, map: map)
        return event
    }

    private func eventSubscriptionInRetryEvent(_ map: [String: Any?]?) -> SubscriptionInRetryEvent {
        let userData = getUserData(map)
        let data = getDataArray(map)

        var event = SubscriptionInRetryEvent(
            data: data,
            userData: userData
        )
        addPredefinedParameters(event: &event, map: map)
        return event
    }

    private func eventConvertedTrialEvent(_ map: [String: Any?]?) -> ConvertedTrialEvent {
        let userData = getUserData(map)
        let data = getDataArray(map)

        var event = ConvertedTrialEvent(
            data: data,
            userData: userData
        )
        addPredefinedParameters(event: &event, map: map)
        return event
    }

    private func eventConvertedOfferEvent(_ map: [String: Any?]?) -> ConvertedOfferEvent {
        let userData = getUserData(map)
        let data = getDataArray(map)

        var event = ConvertedOfferEvent(
            data: data,
            userData: userData
        )
        addPredefinedParameters(event: &event, map: map)
        return event
    }

    private func eventReactivatedSubscriptionEvent(_ map: [String: Any?]?) -> ReactivatedSubscriptionEvent {
        let userData = getUserData(map)
        let data = getDataArray(map)

        var event = ReactivatedSubscriptionEvent(
            data: data,
            userData: userData
        )
        addPredefinedParameters(event: &event, map: map)
        return event
    }

    private func eventRenewedSubscriptionEvent(_ map: [String: Any?]?) -> RenewedSubscriptionEvent {
        let userData = getUserData(map)
        let data = getDataArray(map)

        var event = RenewedSubscriptionEvent(
            data: data,
            userData: userData
        )
        addPredefinedParameters(event: &event, map: map)
        return event
    }

    private func eventConvertedTrialFromRetryEvent(_ map: [String: Any?]?) -> ConvertedTrialFromRetryEvent {
        let userData = getUserData(map)
        let data = getDataArray(map)

        var event = ConvertedTrialFromRetryEvent(
            data: data,
            userData: userData
        )
        addPredefinedParameters(event: &event, map: map)
        return event
    }

    private func eventConvertedOfferFromRetryEvent(_ map: [String: Any?]?) -> ConvertedOfferFromRetryEvent {
        let userData = getUserData(map)
        let data = getDataArray(map)

        var event = ConvertedOfferFromRetryEvent(
            data: data,
            userData: userData
        )
        addPredefinedParameters(event: &event, map: map)
        return event
    }

    private func eventRenewedSubscriptionFromRetryEvent(_ map: [String: Any?]?) -> RenewedSubscriptionFromRetryEvent {
        let userData = getUserData(map)
        let data = getDataArray(map)

        var event = RenewedSubscriptionFromRetryEvent(
            data: data,
            userData: userData
        )
        addPredefinedParameters(event: &event, map: map)
        return event
    }

    private func eventUnsubscriptionEvent(_ map: [String: Any?]?) -> UnsubscriptionEvent {
        let userData = getUserData(map)
        let data = getDataArray(map)

        var event = UnsubscriptionEvent(
            data: data,
            userData: userData
        )
        addPredefinedParameters(event: &event, map: map)
        return event
    }
}
